autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java

```diff
@@ -33,17 +42,26 @@ public FreeIpaCommunicator(UserV1Endpoint userV1Endpoint) {
     public SyncOperationStatus synchronizeAllUsers(SynchronizeAllUsersRequest request) {
         LOGGER.info("Invoking freeIpa user sync request: {}", request);
         String envCrn = request.getEnvironments().iterator().next();
-        SyncOperationStatus lastSyncStatus = ThreadBasedUserCrnProvider.doAsInternalActor(
-                internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
-                () -> userV1Endpoint.getLastSyncOperationStatus(envCrn));
-        if (SynchronizationStatus.RUNNING.equals(lastSyncStatus.getStatus())) {
-            LOGGER.info("There is a user sync operation already running for environment: {} with operationId: {}, " +
-                    "skipping request to trigger another user sync", envCrn, lastSyncStatus.getOperationId());
-            return lastSyncStatus;
-        } else {
-            return ThreadBasedUserCrnProvider.doAsInternalActor(
+        try {
+            SyncOperationStatus lastSyncStatus = ThreadBasedUserCrnProvider.doAsInternalActor(
                     internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
-                    () -> userV1Endpoint.synchronizeAllUsers(request));
+                    () -> userV1Endpoint.getLastSyncOperationStatus(envCrn));
+            if (SynchronizationStatus.RUNNING.equals(lastSyncStatus.getStatus())) {
+                LOGGER.info("There is a user sync operation already running for environment: {} with operationId: {}, " +
+                        "skipping request to trigger another user sync", envCrn, lastSyncStatus.getOperationId());
+                return lastSyncStatus;
+            } else {
+                SyncOperationStatus status = ThreadBasedUserCrnProvider.doAsInternalActor(
+                        internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
+                        () -> userV1Endpoint.synchronizeAllUsers(request));
+                metricService.incrementMetricCounter(IPA_USER_SYNC_INVOCATION, MetricTag.TENANT.name(), request.getAccountId());
+                return status;
+            }
+        } catch (Exception ex) {
+            LOGGER.error("Encountered error when trying to synchronize users to IPA for environment: {}", envCrn, ex);
+            metricService.incrementMetricCounter(IPA_USER_SYNC_FAILED, MetricTag.TENANT.name(), request.getAccountId());
+
+            throw new RuntimeException(format("Failed to perform user sync for environment: %s", envCrn));
```

Comment:
metric will also be incremented even if getLastSyncOperationStatus call fails so it will be treated user sync failed which was actually never attempted

Reply:
Refactored to detect only if IPA user sync failed